### PR TITLE
Support multipart/form-data via http

### PR DESCRIPTION
**Pull request by atsushi-ishibashi**
## Overview
refer: #223 
runbook design.
```
req:
  /upload:
    post:
      body:
        multipart/form-data:
          field1: '/path/to/image1.png'
          field2: '/path/to/image2.png'
```

header and body example
```
//header
Content-Type: multipart/form-data; boundary=1234567890boundary

//body
--1234567890boundary
Content-Disposition: form-data; name="field1"
...
--1234567890boundary
Content-Disposition: form-data; name="field2"
...
--1234567890boundary
```

## Question
Should we use [`CreateFormFile`](https://pkg.go.dev/mime/multipart#Writer.CreateFormFile)? 

The difference in body.
```
//CreateFormField(used currently)
Content-Disposition: form-data; name="field1"

//CreateFormFile
Content-Disposition: form-data; name="field1"; filename="<something input>"
Content-Type: application/octet-stream
```

If we use `CreateFormFile`, we must add `filename` to runbook.

## Todo
When we get a consensus, I'll proceed
- [ ] test
- [ ] docs
- [ ] anything??

### CODE REVIEW

**Comment by k1LoW** [comment]
> Should we use CreateFormFile?

I think it's good to use it.
Customize it when we want to be able to change the Content-Type.

Chrome seems to determine Content-Type from the file.

<details>

### Request example using https://github.com/k1LoW/runn/blob/a67c85db98b03ec6a76e45d8306f90f02285e270/cdp_test.go#L177-L259

**Request Header**

```
map[Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9] Accept-Encoding:[gzip, deflate, br] Accept-Language:[en-US] Cache-Control:[max-age=0] Connection:[keep-alive] Content-Length:[2702] Content-Type:[multipart/form-data; boundary=----WebKitFormBoundaryz7ufvqPCts3EzoQT] Origin:[http://127.0.0.1:59561] Referer:[http://127.0.0.1:59561/form] Sec-Fetch-Dest:[document] Sec-Fetch-Mode:[navigate] Sec-Fetch-Site:[same-origin] Sec-Fetch-User:[?1] Upgrade-Insecure-Requests:[1] User-Agent:[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/107.0.5304.110 Safari/537.36]]
```

**Request Body**

```
------WebKitFormBoundaryz7ufvqPCts3EzoQT
Content-Disposition: form-data; name="username"


------WebKitFormBoundaryz7ufvqPCts3EzoQT
Content-Disposition: form-data; name="upload"; filename="logo.svg"
Content-Type: image/svg+xml

<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 600 700" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" xmlns:serif="http://www.serif.com/" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;">
    <g transform="matrix(0.837956,0,0,0.837956,-109.966,91.2273)">
        <g transform="matrix(1.6932,0,0,0.933057,-266.214,-33.1687)">
            <path d="M383.734,315.248L361.062,290.238L287.591,354.124L310.263,379.134L383.734,315.248Z" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.933057,-141.328,26.2343)">
            <path d="M383.734,315.248L361.062,290.238L287.591,354.124L310.263,379.134L383.734,315.248Z" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.933057,-17.1507,-33.1687)">
            <path d="M383.734,315.248L361.062,290.238L287.591,354.124L310.263,379.134L383.734,315.248Z" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.933057,107.165,-33.1687)">
            <path d="M383.734,315.248L361.062,290.238L287.591,354.124L310.263,379.134L383.734,315.248Z" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.750159,-265.529,100.003)">
            <rect x="287.187" y="183.478" width="23.076" height="189.226" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.750159,-141.042,100.242)">
            <rect x="287.187" y="183.478" width="23.076" height="189.226" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.750159,-16.555,100.003)">
            <rect x="287.187" y="183.478" width="23.076" height="189.226" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.750159,107.932,100.003)">
            <rect x="287.187" y="183.478" width="23.076" height="189.226" style="fill:rgb(36,41,46);"/>
        </g>
        <g transform="matrix(1.6932,0,0,0.750159,232.419,100.003)">
            <rect x="287.187" y="183.478" width="23.076" height="189.226" style="fill:rgb(36,41,46);"/>
        </g>
    </g>
</svg>

------WebKitFormBoundaryz7ufvqPCts3EzoQT
Content-Disposition: form-data; name="submit"

Submit
------WebKitFormBoundaryz7ufvqPCts3EzoQT--

```
</details>

> we must add filename to runbook.

How about getting it from the name of the file to be set?

**Comment by atsushi-ishibashi** [comment]
Ok, I replace CreateFormField with CreateFormFile.

>Chrome seems to determine Content-Type from the file.
I see. About content-type, I check if the below function works well
https://pkg.go.dev/net/http#DetectContentType

**Comment by atsushi-ishibashi** [comment]
encoding doesn't work well. kin-openapi seems to support it [here](https://github.com/getkin/kin-openapi/blob/871e029d62a4d47239d11006431407e08edb1259/openapi3filter/validate_request.go#L247)
https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#encoding-object
```
$ go test -run=TestHttp                                                                                          
--- FAIL: TestHttp (0.01s)
    --- FAIL: TestHttp/testdata/book/http.yml (0.01s)
        operator_test.go:581: failed to run testdata/book/http.yml: http request failed on 'Test using HTTP'.steps.fileupload: openapi3 validation error: request body has an error: failed to decode request body: path file1: unsupported content type "image/png"
            -----START HTTP REQUEST-----
            POST /upload HTTP/1.1
            Host: 127.0.0.1:54660
            Content-Type: multipart/form-data; boundary=18a0f53ce6d7f985368b887a563c469f295cca86758196b9342966422ccc
            
            --18a0f53ce6d7f985368b887a563c469f295cca86758196b9342966422ccc
            Content-Disposition: form-data; name="file1"; filename="dummy.png"
            Content-Type: image/png
```

**Comment by atsushi-ishibashi** [comment]
I got it. Register png and jpeg but https://github.com/k1LoW/runn/pull/283/commits/d6c3cd0c1a9255712456205b18bab3c347b7a5e6 doesn't support others for example svg, webp. How do we consider mime type coverage?
https://github.com/getkin/kin-openapi/blob/871e029d62a4d47239d11006431407e08edb1259/openapi3filter/req_resp_decoder.go#L911

**Comment by atsushi-ishibashi** [comment]
Btw `make test` stucks. Maybe around Google Chrome Helper(Renderer). Do you know how to recover?

**Comment by k1LoW** [comment]
> make test stucks. Maybe around Google Chrome Helper(Renderer). Do you know how to recover?

Now I force kill chrome if I have any problems.

Maybe we should implement a default timeout in the CDP Runner.

**Comment by k1LoW** [comment]
> How do we consider mime type coverage?

It would be good to add a mime-type when it becomes necessary.

(Eventually, we may get a list of file mime-types from somewhere for coverage)

**Comment by k1LoW** [comment]
@atsushi-ishibashi 

> make test stucks.

Does this occur every time?

**Comment by atsushi-ishibashi** [comment]
> Does this occur every time?

yes, every time. I try after killing chrome but it still stuck😢

**Comment by atsushi-ishibashi** [comment]
I installed chrome dev version and canary version in my laptop so it may cause

**Comment by atsushi-ishibashi** [comment]
I couldn't find good library to fetch mime type so hard code refer to https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types

**Comment by atsushi-ishibashi** [comment]
Btw `make` works well, thanks

**Comment by atsushi-ishibashi** [comment]
For the golden test, I can enable you to set fixed bounday. Welcome to better env key than `TEST_MODE`.
https://github.com/k1LoW/runn/blob/b55157794df52a5ff00038a60abd5390732e7e4b/http.go#L149

**Comment by atsushi-ishibashi** [comment]
Can I leave you to update golden file after your review? I think that contributors should not update golden files👀
If you allow me to do, I will

**Comment by k1LoW** [comment]
> Can I leave you to update golden file after your review? I think that contributors should not update golden files👀

OK.

However, you can update the existing golden file without any concern.
Because we can review that golden file as well.

To update the golden file, run `env UPDATE_GOLDEN=1 make test`.

**Comment by atsushi-ishibashi** [comment]
>Because we can review that golden file as well

That's right. I will

**Comment by atsushi-ishibashi** [comment]
I don't understand why `TestSetUploadFile` fail. please give me hint🙋‍♂️ 
```
--- FAIL: TestSetUploadFile (30.03s)
    cdp_test.go:225: context canceled
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 81 [running]:
testing.tRunner.func1.2({0x100e0a2e0, 0xc0000464b0})
	/usr/local/Cellar/go/1.19.3/libexec/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/usr/local/Cellar/go/1.19.3/libexec/src/testing/testing.go:1399 +0x39f
panic({0x100e0a2e0, 0xc0000464b0})
	/usr/local/Cellar/go/1.19.3/libexec/src/runtime/panic.go:884 +0x212
github.com/k1LoW/runn.TestSetUploadFile(0xc0005831e0)
	.../runn/cdp_test.go:230 +0x945
```

TestRunnable seems to encode multi part body as `(map[string]interface{})`. There is a big difference body from yaml config so I wonder how we solve this...
```
--- FAIL: TestRunnable (0.09s)
    --- FAIL: TestRunnable/http.yml (0.01s)
        runbook_test.go:106: failed to run /var/folders/08/5pkbt8xn2ys_gbk9jqbqgc0w0000gp/T/TestRunnablehttp.yml248152161/001/-testdata-book-http.yml: http request failed on 'Captured of http.yml run'.steps[2]: invalid body1: --f0e3c453f737b07f185303ec08c0321d05a9ff9b12d173727cbb49d38904\r\nContent-Disposition: form-data; name="file2"; filename="dummy.jpeg"\r\nContent-Type: image/jpeg\r\n\r\n\\\\\\JFIF\\\\\`\`\\\\\>CREATOR: gd-jpeg v1.0 (using IJG JPEG v80), default quality\n\\\C\\b\\\a\\\b\a\a\a\t\t\b\n\f\\r\f\v\v\f\\\\\\\\\\\\\ $.' ",#\\(7),01444\'9=82<.342\\\C\\t\t\t\f\v\f\\r\r\2!\!22222222222222222222222222222222222222222222222222\\\\\b\(\<\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\a\b\t\n\v\\\\\\\\\\\\\\\\\\\\}\\\\\\\\!1A\\Qa\a"q\2\\\\b#B\\\R\\$3br\\t\n\\\\\%&'()*456789:CDEFGHIJSTUVWXYZcdefghijstuvwxyz\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\a\b\t\n\v\\\\\\\\\\\\\\a\\\\\\w\\\\\\\!1\\AQ\aaq\"2\\b\B\\\\\t#3R\\br\\n\$4\%\\\\\&'()*56789:CDEFGHIJSTUVWXYZcdefghijstuvwxyz\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\f\\\\\\\\?\\\\(\\\(\\\(\\\(\\\(\\\T\h\\B\H\\@g\\\\\\\\f\\yݖ屌5e\:l\\\\\\\U-\0s\\q\\\N\\2\\\M9t\\u\\\\\H\bm\\\\\\\&y\Rx<f\AD\\\\\\\y;vn\\6\\wnq\5N\\(\\\(\\\(\\\(\\\(\\\\\\r\n--f0e3c453f737b07f185303ec08c0321d05a9ff9b12d173727cbb49d38904\r\nContent-Disposition: form-data; name="file1"; filename="dummy.png"\r\nContent-Type: image/png\r\n\r\n\PNG\r\n\\n\\\\rIHDR\\\<\\\(\\\\\P\\\\\\\PLTE\\\\\\___\\\???\\\\\\\\\\\.C\\\\tpHYs\\\\\\\\\\+\\\\\aIDAT8\c`\\\\\0)1)\\,\nP\\n0aPfv`f`f`R\20\U\\\\\\\\\\\\f\f\\\\\\9\\\Vi\\\\\*@\\\\\\\j\\\\`\fTȪJ\\F\\b\\\\\a\\ز\\\\\IEND\B`\\r\n--f0e3c453f737b07f185303ec08c0321d05a9ff9b12d173727cbb49d38904--\r\n
FAIL
```

File: testutil/http.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ func HTTPServerAndRouter(t *testing.T) (*httptest.Server, *httpstub.Router) {
 	r := httpstub.NewRouter(t)
 	r.Method(http.MethodPost).Path("/users").Response(http.StatusCreated, nil)
 	r.Method(http.MethodPost).Path("/help").Response(http.StatusCreated, nil)
+	r.Method(http.MethodPost).Path("/upload").Response(http.StatusCreated, nil)
```
Comment (k1LoW) [inline testutil/http.go:40]: @atsushi-ishibashi 

About https://github.com/k1LoW/runn/pull/283#issuecomment-1316396837 ,

I think another test (`TestSetUploadFile`) is failing because of the duplicate `/upload` path.

https://github.com/k1LoW/runn/pull/283/files#diff-d833779323fc4c6d7f044dc910cb45153e44fb79b85e01d39d77d6aa8a6f9b67R53

**Comment by k1LoW** [comment]
> TestRunnable seems to encode multi part body as (map[string]interface{}). There is a big difference body from yaml config so I wonder how we solve this...

How about creating a new runbook and testing only file uploads? ( and revert http.yml. )

File: testdata/http_multipart.yml.runbook.golden [hunk 2]
```diff
@@ -0,0 +1,43 @@
+-- -testdata-book-http_multipart.yml --
+desc: Captured of http_multipart.yml run
+runners:
+  req: '[THIS IS HTTP RUNNER]'
+steps:
+- req:
+    /multipartUpload:
+      post:
+        body:
+          multipart/form-data; boundary=123456789012345678901234567890abcdefghijklmnopqrstuvwxyz: !!binary |
```
Comment (k1LoW) [inline testdata/http_multipart.yml.runbook.golden:10]: fmm... There seems to be a bug in https://github.com/k1LoW/runn/blob/main/capture/runbook.go.
Content-Type header is not being retrieved.

**Review by k1LoW — COMMENTED** [review]
Is it possible to compare uploaded files as well as https://github.com/k1LoW/runn/blob/3d14d1d7d09e3dabd600ba60d8a8e536841ffeff/cdp_test.go#L177 ?

`testutil.HTTPServerAndRouter(t)` can retrieve received requests in sequence (using https://github.com/k1LoW/httpstub).

https://github.com/k1LoW/runn/blob/3d14d1d7d09e3dabd600ba60d8a8e536841ffeff/cdp_test.go#L239

File: testdata/http_multipart.yml.runbook.golden (hunk 2 above)
Comment (k1LoW) [reply testdata/http_multipart.yml.runbook.golden:10]: I'll try to fix this bug with another PR on my part.

**Comment by k1LoW** [comment]
https://github.com/k1LoW/runn/pull/283#pullrequestreview-1182166701

If it is hard, I will add it myself 👍

File: http.go [hunk 3]
```diff
@@ -117,6 +130,59 @@ func (r *httpRequest) encodeBody() (io.Reader, error) {
 	}
 }
 
+func (r httpRequest) isMultipartFormDataMediaType() bool {
```
Comment (k1LoW) [inline http.go:133]: need pointer?

```suggestion
func (r *httpRequest) isMultipartFormDataMediaType() bool {
```

File: http.go [hunk 4]
```diff
@@ -117,6 +130,59 @@ func (r *httpRequest) encodeBody() (io.Reader, error) {
...
+
+func (r *httpRequest) encodeMultipart() (io.Reader, error) {
+	quoteEscaper := strings.NewReplacer("\\", "\\\\", `"`, "\\\"")
+	values, ok := r.body.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("invalid body: %v", r.body)
+	}
+	buf := &bytes.Buffer{}
+	mw := multipart.NewWriter(buf)
+	if os.Getenv("TEST_MODE") == "true" {
```
Comment (k1LoW) [inline http.go:148]: How about creating `httpRequest.boundary` and injecting values as needed?

File: http.go (hunk 3 above)
Comment (atsushi-ishibashi) [reply http.go:133]: no need pointer because this is not mutable method🙋 If it is convention, I follow you👍

Comment (k1LoW) [reply http.go:133]: > no need pointer because this is not mutable method

You are certainly right.

> If it is convention, I follow you

It is not convention. I had made it a pointer unconditionally, so I learned a lot. Thank you!

File: testdata/book/http_multipart.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    desc: Post /multipartUpload
+    req:
+      /multipartUpload:
+        post:
+          body:
+            multipart/form-data:
+              file1: "{{ vars.baseDir }}/testdata/dummy.png"
+              file2: "{{ vars.baseDir }}/testdata/dummy.jpeg"
+    test: |
+      current.res.status == 201
```
[the remark this answers is not in the document]
Comment (atsushi-ishibashi) [reply testdata/book/http_multipart.yml]: fix b1ffe1a

File: http.go [hunk 6]
```diff
@@ -117,6 +130,59 @@ func (r *httpRequest) encodeBody() (io.Reader, error) {
...
+		_ = mw.SetBoundary("123456789012345678901234567890abcdefghijklmnopqrstuvwxyz")
+	}
+	for fieldName, ifileName := range values {
+		fileName, ok := ifileName.(string)
+		if !ok {
+			return nil, fmt.Errorf("invalid body: %v", r.body)
+		}
+		fileBody, err := os.ReadFile(filepath.Clean(fileName))
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (atsushi-ishibashi) [reply http.go]: fix b1ffe1a
btw I missed `TestRequestBody` check😅

**Comment by atsushi-ishibashi** [comment]
> [#283 (review)](https://github.com/k1LoW/runn/pull/283#pullrequestreview-1182166701)
> 
> If it is hard, I will add it myself 👍

I try but when I find it is hard for me, let you know!

File: http.go (hunk 4 above)
Comment (atsushi-ishibashi) [reply http.go:148]: Is this what you mean? I'd like to fix boundary in `TestRunbook` so I cannot access `httpRequest`.
```
type httpRequest struct {
	path      string
	method    string
	headers   map[string]string
	mediaType string
	body      interface{}

	multipartWriter *multipart.Writer
	boundary        string
}
...
buf := &bytes.Buffer{}
mw := multipart.NewWriter(buf)
if r.boundary != "" {
  _ = mw.SetBoundary(r.boundary)
  //_ = mw.SetBoundary("123456789012345678901234567890abcdefghijklmnopqrstuvwxyz")
}
```

Or what you mean is supporting boundary via yaml book?
```
req:
  /upload:
    post:
      body:
        multipart/form-data:
          boundary: 1234567890abcdefg
          field1: '/path/to/image1.png'
          field2: '/path/to/image2.png'
```

File: testdata/book/http_multipart.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (atsushi-ishibashi) [reply testdata/book/http_multipart.yml]: uh, not fix...

File: http.go (hunk 4 above)
Comment (k1LoW) [reply http.go:148]: > in TestRunbook so I cannot access httpRequest

Oh, I see.

To inject boundary, we need to pass the value via HTTP Runner with httpRunnerOption.

I'll try to fix it at another time. Thank you 👍

File: testdata/http_multipart.yml.runbook.golden (hunk 2 above)
Comment (k1LoW) [reply testdata/http_multipart.yml.runbook.golden:10]: Ah, Content-Type `multipart/form-data; boundary=123456789012345678901234567890abcdefghijklmnopqrstuvwxyz:` is getting it.

**Comment by atsushi-ishibashi** [comment]
>Is it possible to compare uploaded files as well as

How about the way of a778a57 ?

**Comment by k1LoW** [comment]
> How about the way of https://github.com/k1LoW/runn/commit/a778a57dd9c6bbe36bd5b1e29811fafe07bc60d1 ? I need to setenv so separate test case.

My suggestion was to check if the request created is uploading the file correctly.
One way to do this is to compare the file to be uploaded with the file received by the test server.

https://github.com/k1LoW/runn/blob/3d14d1d7d09e3dabd600ba60d8a8e536841ffeff/cdp_test.go#L247-L257

**Comment by atsushi-ishibashi** [comment]
>One way to do this is to compare the file to be uploaded with the file received by the test server.

Ok, I will in a new test case.

Btw `TestRunbook` fail randomly because of undetermined order.
```
multipart/form-data:
  file1: "{{ vars.baseDir }}/testdata/dummy.png"
  file2: "{{ vars.baseDir }}/testdata/dummy.jpeg"
  name: bob
```

For user, array is better?
```
multipart/form-data:
  - file1: "{{ vars.baseDir }}/testdata/dummy.png"
  - file2: "{{ vars.baseDir }}/testdata/dummy.jpeg"
  - name: bob
```

File: capture/runbook_test.go [hunk 7]
```diff
@@ -17,13 +17,16 @@ func TestRunbook(t *testing.T) {
 		book string
 	}{
 		{filepath.Join(testutil.Testdata(), "book", "http.yml")},
+		{filepath.Join(testutil.Testdata(), "book", "http_multipart.yml")},
```
Comment (k1LoW) [inline capture/runbook_test.go:20]: https://github.com/k1LoW/runn/pull/283#issuecomment-1318150788

We think `multipart/form-data`  and `application/x-www-form-urlencoded` should be compatible as YAML formats.
So we think it is fine to leave it as it is now.

TestRunbook's problem is caused by a problem in capture/runbook.go, so I think this line can be removed.

Comment (k1LoW) [reply capture/runbook_test.go:20]: ( I think it probably needs to parse multipart/form-data request body at capture/runbook.go. ref: https://github.com/k1LoW/runn/pull/283#discussion_r1024579919)

File: capture/runbook_test.go [hunk 8]
```diff
@@ -17,13 +17,17 @@ func TestRunbook(t *testing.T) {
 		book string
 	}{
 		{filepath.Join(testutil.Testdata(), "book", "http.yml")},
+		// FIXME: https://github.com/k1LoW/runn/pull/283#discussion_r1024812525
```
Comment (k1LoW) [inline capture/runbook_test.go:20]: 💪

**Review by k1LoW — APPROVED** [review]
@atsushi-ishibashi Thank you for your contribution!!!

( I will try to fix the remaining issues after this PR is merged. )
